benchmarks/libseff/generator/main.c

```diff
@@ -0,0 +1,142 @@
+#include "seff.h"
+#include <stdio.h>
+
+DEFINE_EFFECT(yield, 0, void, { int64_t value; });
+
+// Tree data structure
+struct node_t {
+  struct node_t* l;
+  int64_t v;
+  struct node_t* r;
+};
+typedef struct node_t node_t;
+typedef struct node_t* tree_t;
+
+// Generator data structure
+struct thunk_t {
+  int64_t v;
+  struct thunk_t* next;
+};
+typedef struct thunk_t thunk_t;
+typedef thunk_t* generator_t;
+
+// Allocate and free a tree node
+static inline node_t* allocNode() { return malloc(sizeof(node_t));}
+static inline void freeNode(node_t* node) { free(node); }
+
+// Allocate a tree on the heap
+static tree_t makeTree(int n) {
+  if (n == 0) return NULL;
+  tree_t node = allocNode();
+  tree_t t = makeTree (n - 1);
+
+  // Note: both left and right pointers point to the same memory, 
+  // to match the other languages
+  node->l = t;
+  node->r = t;
+  node->v = n;
+  return node;
+}
+
+// Free tree from heap
+static void freeTree(tree_t tree) {
+  if (tree == NULL) return;
+  freeTree(tree->l);
+  freeNode(tree);
+}
+
+// Allocate and free generator thunk
+static inline thunk_t* allocThunk() { return malloc(sizeof(thunk_t)); }
+static inline void freeThunk(thunk_t* t) { free(t); }
+
+// Free generator memory
+static void freeGenerator(generator_t generator) {
+  generator_t next;
+  while (generator != NULL) {
+    next = generator->next; 
+    freeThunk(generator);
+    generator = next;
+  }
+}
+
+// Yield handler
+static generator_t handleYield(seff_coroutine_t *k) {
+  effect_set handles_yield = HANDLES(yield); 
+  seff_request_t req = seff_handle(k, NULL, handles_yield);
+
+  generator_t generator = allocThunk(); 
+  generator->v = -1;
+  generator_t currentThunk = generator;
+  bool done = false;
+  while (!done) {   
+    switch (req.effect) {
+      CASE_EFFECT(req, yield, { 
+        currentThunk->v = payload.value;
+        currentThunk->next = allocThunk();
+        currentThunk = currentThunk->next;
+        req = seff_handle(k, 0, handles_yield);
+        break;
+      })
+      CASE_RETURN(req, {
+        currentThunk->next = NULL; // Empty thunk 
+        done = true;
+        break;
+      })
+    };
+  }
```

Comment:
I think this algorithm for creating a generator is a bit unfair with `libseff`, since it materializes the whole generator into a linked list (`thunk_t`) therefore consuming memory linear with the size of the tree.
Whereas the OCaml one returns the next value and a closure to get the next one.

Since closures in C are a bit un-idiomatic I'd change this to a tuple of the value and the coroutine that it needs to execute next.

```c
typedef struct {
  int64_t v;
  seff_coroutine_t *k;
} generator_t;
```
And then sum could do something like:

```c
  while (generator = handleYield(generator.k), generator.k != NULL) {
    acc += generator.v;
  }
```

I've tested it quickly locally and it also improves performance quite a lot.

Reply:
Good point! I will make the change.

Reply:
I've added the change, let me know if this is what you had in mind!

Reply:
Yup, thank you!